content/en/docs/Responses/Child.md

```diff
@@ -11,30 +11,95 @@ description: >
 {{< tab header="**Example**:" disabled=true />}}
 {{< tab header="OpenSubsonic" lang="json">}}
 {
-  "id": "082f435a363c32c57d5edb6a678a28d4",
-  "parent": "e8a0685e3f3ec6f251649af2b58b8617",
-  "isDir": false,
-  "title": "\"polar expedition\"",
-  "album": "Live at The Casbah - 2005-04-29",
-  "artist": "The New Deal",
-  "track": 4,
-  "year": 2005,
-  "coverArt": "mf-082f435a363c32c57d5edb6a678a28d4_6410b3ce",
-  "size": 19866778,
-  "contentType": "audio/flac",
-  "suffix": "flac",
-  "starred": "2023-03-27T09:45:27Z",
-  "duration": 178,
-  "bitRate": 880,
-  "path": "The New Deal/Live at The Casbah - 2005-04-29/04 - \"polar expedition\".flac",
-  "playCount": 8,
-  "played": "2023-03-26T22:27:46Z",
-  "discNumber": 1,
-  "created": "2023-03-14T17:51:22.112827504Z",
-  "albumId": "e8a0685e3f3ec6f251649af2b58b8617",
-  "artistId": "97e0398acf63f9fb930d7d4ce209a52b",
-  "type": "music",
-  "isVideo": false
+    "id": "082f435a363c32c57d5edb6a678a28d4",
+    "parent": "e8a0685e3f3ec6f251649af2b58b8617",
+    "isDir": false,
+    "title": "\"polar expedition\"",
+    "album": "Live at The Casbah - 2005-04-29",
+    "artist": "The New Deal",
+    "track": 4,
+    "year": 2005,
+    "coverArt": "mf-082f435a363c32c57d5edb6a678a28d4_6410b3ce",
+    "size": 19866778,
+    "contentType": "audio/flac",
+    "suffix": "flac",
+    "starred": "2023-03-27T09:45:27Z",
+    "duration": 178,
+    "bitRate": 880,
+    "path": "The New Deal/Live at The Casbah - 2005-04-29/04 - \"polar expedition\".flac",
+    "playCount": 8,
+    "played": "2023-03-26T22:27:46Z",
+    "discNumber": 1,
+    "created": "2023-03-14T17:51:22.112827504Z",
+    "albumId": "e8a0685e3f3ec6f251649af2b58b8617",
+    "artistId": "97e0398acf63f9fb930d7d4ce209a52b",
+    "type": "music",
+    "isVideo": false,
+    "bpm": 134,
+    "comment": "This is a song comment",
+    "sortName": "Polar expedition",
+    "musicBrainzId": "189002e7-3285-4e2e-92a3-7f6c30d407a2",
+    "genres": [
+        "Hip-Hop",
+        "East coast"
+    ],
+    "artists": [
+        {
+            "id": "ar-1",
+            "name": "Artist 1"
+        },
+        {
+            "id": "ar-2",
+            "name": "Artist 2"
+        }
+    ],
+    "displayArtist": "Artist 1 feat. Artist 2",
+    "albumArtists": [
+        {
+            "id": "ar-6",
+            "name": "Artist 6"
+        },
+        {
+            "id": "ar-7",
+            "name": "Artist 7"
+        }
+    ],
+    "displayAlbumArtist": "Artist 6 & Artist 7",
+    "contributors": [
+        {
+            "role": "composer",
```

Comment:
what about "performer" with its subroles? (like drums, bass guitar, etc.) will see this later?

Reply:
Never saw that on any provider, those are just roles like others no?
We can define a role subtype if necessary but it would simpler to maybe just define those as "performer/bass" role ?

Reply:
This is yet another example of messy fields in tags (see https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html#performer to see what picard does, according to file formats).
Again, this is still better than no field at all. But I mean, a server maintains the music library and should try hard to support a lot of format specificities and unify everything in its database (this is basically reconstructing the database used to tag the files). And the API should reflect that. If we just provide raw data that are read in tags and make all the server/client chain ignore their contents, I think we miss some presentation opportunities.
For example, the "credits" view of Roon: https://help.roonlabs.com/galleryDocuments/edbsn03b66b1efc4daf4a0b94d8ddaa04209a045cf8d85fb2d5879c34a0abb5510bf4d27ec0bf4c3a3c786f1e826116971ca8?inline=true
This is still possible with this proposal, but each client will have to further reparse role things, make guesses, etc.. At least, let's put an example with an instrument and/or chose a format that may simplify this.
(same for release types, we can present like this: https://help.roonlabs.com/galleryDocuments/edbsn6b5862c5f4c475b9b28dd151a3fa9f6100ecb1f9d69a41aed61af4666001672c17eb399b05eaa0555c095632270e3e5c?inline=true (see the albums/singles tabs), this is why I consider this field "important" and not just an info that users can build filters on)

Reply:
The thing is that there's no magic on the servers, if the user have bad tags, the server can't normalize it, and if the user have proper tags then there's nothing to be done by the server. This is a tag software / user issue here more than anything. Not all servers use MB and not all users want to use MB anyway.

So we can put some recommendation in the docs to help users and servers that use MB data, but we can't enforce anything. Enforcing means that any data that does not fit the cases we expose can't be exposed and this breaks the liberty of the users to tag as they want to workaround some software limitations.

You have no idea how creative users are with the tags to sort their media via the wrongs tags because an app support that tag and not another.

With the roles you can already do the roon style thing, and the release value too.

So since you handle MB data that override tags, propose the format of the release and role field for subtypes and I'll add notes to the doc.

Reply:
For roles I would say:
* "role", string
* "subroles", array of strings


For release types:
* "releaseType", string
* "secondaryReleaseTypes", array of strings

Reply:
What I meant was a format in a single string :) The subparts are only relevant with MB are we sure we want to force that road to all?
Tags are supported by all servers, MB by nearly none so far.

Reply:
I understand your point, but mine is that if we want to aim to improve metadata quality, it would be better to introduce less open/clunky fields. Otherwise I am afraid they will be not that useful and likely difficult to modify afterwards.
But fine, let's just put a "role", that's better than nothing.

Reply:
The problem is what do the server that only have 1 tag single string and need to fill those arrays that they can't? In all cases this will generate differences between servers.
I'm all in for more data, but we can't force servers to use MB. Most don't and won't, needs API keys, finding a working server, dealing with rate limits, ..... If there's an official way to tag the fields that servers can use to split the data and fill those arrays then I'm OK to add the arrays. If not we can only go with recommendation in a single string and then no arrays.
If the arrays are 100% specific to MB then I don't see the need and it's better to have a simple solution for you, a convention that role is performer/XXXX with XXX being the bass or whatever that client can use. Same for album release Album/XXXX for the sub types.
That covers your need + cover the users who already tag the media like that or they can edit their tags to match that. And this does not force the servers to handle MB while still allowing users to use the feature via the simple tags.

This covers all the cases without blocking people that would prefer to not support this than having to deal with tag normalisation and MB servers.

Reply:
Not sure why you consider this to be MB specific?
Not sure why "performer/XX,YY,ZZZ" would be better than another secondary field (except it brings yet again another questionable separator in a string). If a server does not have sub roles, then it just provides an empty array and that's all?
Not sure why it would block people to just put an empty array for this field?

As I said, the goal of all this is to make clients better. If no client is interesting in this, then this is fine, let's just put a "role", that's better than nothing.

Reply:
You need to think from the other perspective.

At the start there's media files with tags.
releaseType and contributors are relatively well defined and are not multi valued.

So a user can only put a string in those fields and a server without MB can only read that string.

Now we have 2 possibility:
1) We expose on the API 1 string and we define a pattern that users can use.
This allows users and clients to properly work with that scheme without requiring the servers to try to normalize the data and be able to extract those sub values. (The only way would be that they do some separation on a specific separator, sounds familiar?)

2) We expose on the API 1 string and 1 array.
Servers now can either pass the data directly in the 1 string, or do some magic and pass it to the array sometimes, or the server have MB access and pass the data properly.

This means that on the client side, we still need to handle the 1 string + parsing of the well known cases + handling the array.
This brings absolutely no benefits to anyone in the end, just more works for the clients, when the only server that support MB can generate the string with the proper separator easily.

___

releaseType: "XXX" or "XXX/YYYY" or "XXXX/YYY/ZZZZ" 
Covers all needs if the clients support the separator he separates, first is type, rest are subtypes. Works via tags, via MB and without imposing anything impossible to do on the server side. We are normalizing the tags to align with MB it's a win win.

role: "XXXX" or "XXXX/YYYY" (There's never a case for more, the solution is to expose more roles)
Again this covers all the cases directly from tags, allows tags and clients to directly work together without forcing the server to do impossible things when it can't and still working with MB.

`{ "role" : "performer/Bass" , "artist": { .... }, "role" : "performer/leader" : "artist" : { ... } }`
Same artist can have multiple roles, using / as a separator is less an issue because the roles are known and there's none with a / it's not like we need to deal with AC/DC. 

So TL;DR; I do not think the arrays brings much and would only make the life of users more complicated to support advanced stuff if their server do not have MB access.

Reply:
Adding on one point of agreement, even if an artist is in multiple separate roles (for example, composer and vocals), would it not make sense for those to be separate entries? As far as I can tell (from Picard anyway), the role tags are `role:artist(s)`, so having multiple `{ "role": "...", "artist": "same" }` would reflect the tags as well

Reply:
Yes they will be different entries, an artist can have multiple roles, but 1 role entry only contains one artist for simplicity, so if there's 3 composers there's 3 entries with role composer not 1 with 3 artists in it.
This simplify things and will better reflect answers when we add support for role filtering.